File: plsql-checks-testkit/src/main/java/org/sonar/plsqlopen/checks/verifier/PlSqlCheckVerifier.java
/*
 * Sonar PL/SQL Plugin (Community)
 * Copyright (C) 2015-2017 Felipe Zorzo
 * mailto:felipebzorzo AT gmail DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plsqlopen.checks.verifier;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.Assert.fail;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;

import org.sonar.api.batch.fs.internal.DefaultInputFile;
import org.sonar.api.batch.fs.internal.TestInputFileBuilder;
import org.sonar.api.batch.sensor.internal.SensorContextTester;
import org.sonar.api.issue.NoSonarFilter;
import org.sonar.plsqlopen.AnalyzerMessage;
import org.sonar.plsqlopen.checks.PlSqlCheck;
import org.sonar.plsqlopen.metadata.FormsMetadata;
import org.sonar.plsqlopen.squid.PlSqlAstScanner;

import com.google.common.base.Splitter;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Ordering;
import com.google.common.io.Files;
import com.sonar.sslr.api.Token;
import com.sonar.sslr.api.Trivia;

public class PlSqlCheckVerifier extends PlSqlCheck {

    private List<TestIssue> expectedIssues = new ArrayList<>();
    
    public static void verify(String filename, PlSqlCheck check) {
        verify(filename, check, null);
    }
    
    public static void verify(String filename, PlSqlCheck check, FormsMetadata metadata) {
        File file = new File(filename);
        PlSqlCheckVerifier verifier = new PlSqlCheckVerifier();
        
        DefaultInputFile inputFile;
        
        try {
            inputFile = new TestInputFileBuilder("key", filename)
                    .setLanguage("plsqlopen")
                    .setCharset(StandardCharsets.UTF_8)
                    .initMetadata(Files.toString(file, StandardCharsets.UTF_8))
                    .setModuleBaseDir(Paths.get(""))
                    .build();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        
        SensorContextTester context = SensorContextTester.create(new File("."));
        context.fileSystem().add(inputFile);
        
        PlSqlAstScanner scanner = new PlSqlAstScanner(context, ImmutableList.of(check, verifier), new NoSonarFilter(), metadata, false);
        Collection<AnalyzerMessage> issues = scanner.scanFile(inputFile);
        
        Iterator<AnalyzerMessage> actualIssues = getActualIssues(issues);
        List<TestIssue> expectedIssues = Ordering.natural().onResultOf(TestIssue::line).sortedCopy(verifier.expectedIssues);

        for (TestIssue expected : expectedIssues) {
            if (actualIssues.hasNext()) {
                verifyIssue(expected, actualIssues.next());
            } else {
                throw new AssertionError("Missing issue at line " + expected.line());
            }
        }

        if (actualIssues.hasNext()) {
            AnalyzerMessage issue = actualIssues.next();
            throw new AssertionError(
                    "Unexpected issue at line " + line(issue) + ": \"" + issue.getText(Locale.ENGLISH) + "\"");
        }

    }

    private static int line(AnalyzerMessage issue) {
        return issue.getLine();
    }

    private static void verifyIssue(TestIssue expected, AnalyzerMessage actual) {
        if (line(actual) > expected.line()) {
            fail("Missing issue at line " + expected.line());
        }
        if (line(actual) < expected.line()) {
            fail("Unexpected issue at line " + line(actual) + ": \"" + actual.getText(Locale.ENGLISH) + "\"");
        }
        if ((expected.startColumn() != null || expected.endColumn() != null || expected.endLine() != null) && actual.getLocation() == null) {
            fail("The current rule does not support precise location.");
        }
        if (expected.message() != null) {
            assertThat(actual.getText(Locale.ENGLISH)).as("Bad message at line " + expected.line())
                    .isEqualTo(expected.message());
        }
        if (expected.effortToFix() != null) {
            assertThat(actual.getCost()).as("Bad effortToFix at line " + expected.line())
                    .isEqualTo(expected.effortToFix());
        }
        if (expected.startColumn() != null) {
            assertThat(actual.getLocation().startCharacter + 1).as("Bad start column at line " + expected.line())
                    .isEqualTo(expected.startColumn());
        }
        if (expected.endColumn() != null) {
            assertThat(actual.getLocation().endCharacter + 1).as("Bad end column at line " + expected.line())
                    .isEqualTo(expected.endColumn());
        }
        if (expected.endLine() != null) {
            assertThat(actual.getLocation().endLine).as("Bad end line at line " + expected.line())
                    .isEqualTo(expected.endLine());
        }
        if (expected.secondaryLines() != null) {
            assertThat(secondary(actual)).as("Bad secondary locations at line " + expected.line())
                    .isEqualTo(expected.secondaryLines());
        }
    }

    private static List<Integer> secondary(AnalyzerMessage issue) {
        List<Integer> result = new ArrayList<>();

        for (AnalyzerMessage issueLocation : issue.getSecondaryLocations()) {
            result.add(issueLocation.getLine());
        }

        return Ordering.natural().sortedCopy(result);
    }

    private static Iterator<AnalyzerMessage> getActualIssues(Collection<AnalyzerMessage> issues) {
        List<AnalyzerMessage> sortedIssues = Ordering.natural().onResultOf(PlSqlCheckVerifier::line).sortedCopy(issues);
        return sortedIssues.iterator();
    }

    @Override
    public void visitToken(Token token) {
        for (Trivia trivia : token.getTrivia()) {
            String text = trivia.getToken().getValue().substring(2).trim();
            String marker = "Noncompliant";

            if (text.startsWith(marker)) {
                int issueLine = trivia.getToken().getLine();
                String paramsAndMessage = text.substring(marker.length()).trim();

                if (paramsAndMessage.startsWith("@")) {
                    String[] spaceSplit = paramsAndMessage.split("[\\s\\[{]", 2);
                    
                    String shiftValue = spaceSplit[0];
                    
                    if (shiftValue.charAt(1) != '+' && shiftValue.charAt(1) != '-') {
                        fail("Use only '@+N' or '@-N' to shifts messages.");
                    }
                    
                    issueLine += Integer.valueOf(shiftValue.substring(1));
                    paramsAndMessage = spaceSplit.length > 1 ? spaceSplit[1] : "";
                }

                TestIssue issue = TestIssue.create(null, issueLine);
                
                if (paramsAndMessage.startsWith("{{")) {
                    int endIndex = paramsAndMessage.indexOf("}}");
                    String message = paramsAndMessage.substring(2, endIndex);
                    issue.message(message);
                    paramsAndMessage = paramsAndMessage.substring(endIndex + 2).trim();
                }
                
                if (paramsAndMessage.startsWith("[[")) {
                    int endIndex = paramsAndMessage.indexOf("]]");
                    addParams(issue, paramsAndMessage.substring(2, endIndex));
                }

                expectedIssues.add(issue);

            } else if (text.startsWith("^")) {
                addPreciseLocation(trivia);
            }
        }
    }

    private static void addParams(TestIssue issue, String params) {
        for (String param : Splitter.on(';').split(params)) {
            int equalIndex = param.indexOf('=');
            if (equalIndex == -1) {
                throw new IllegalStateException("Invalid param at line 1: " + param);
            }
            String name = param.substring(0, equalIndex);
            String value = param.substring(equalIndex + 1);

            if ("effortToFix".equalsIgnoreCase(name)) {
                issue.effortToFix(Integer.valueOf(value));

            } else if ("sc".equalsIgnoreCase(name)) {
                issue.startColumn(Integer.valueOf(value));

            } else if ("ec".equalsIgnoreCase(name)) {
                issue.endColumn(Integer.valueOf(value));

            } else if ("el".equalsIgnoreCase(name)) {
                issue.endLine(lineValue(issue.line(), value));

            } else if ("secondary".equalsIgnoreCase(name)) {
                addSecondaryLines(issue, value);

            } else {
                throw new IllegalStateException("Invalid param at line 1: " + name);
            }
        }
    }

    private static void addSecondaryLines(TestIssue issue, String value) {
        List<Integer> secondaryLines = new ArrayList<>();
        if (!"".equals(value)) {
            for (String secondary : Splitter.on(',').split(value)) {
                secondaryLines.add(lineValue(issue.line(), secondary));
            }
        }
        issue.secondary(secondaryLines);
    }

    private static int lineValue(int baseLine, String shift) {
        if (shift.startsWith("+")) {
            return baseLine + Integer.valueOf(shift.substring(1));
        }
        if (shift.startsWith("-")) {
            return baseLine - Integer.valueOf(shift.substring(1));
        }
        return Integer.valueOf(shift);
    }

    private void addPreciseLocation(Trivia trivia) {
        Token token = trivia.getToken();
        int line = token.getLine();
        String text = token.getValue();
        if (token.getColumn() > 1) {
            throw new IllegalStateException(
                    "Line " + line + ": comments asserting a precise location should start at column 1");
        }
        String missingAssertionMessage = String.format(
                "Invalid test file: a precise location is provided at line %s but no issue is asserted at line %s",
                line, line - 1);
        if (expectedIssues.isEmpty()) {
            throw new IllegalStateException(missingAssertionMessage);
        }
        TestIssue issue = expectedIssues.get(expectedIssues.size() - 1);
        if (issue.line() != line - 1) {
            throw new IllegalStateException(missingAssertionMessage);
        }
        issue.endLine(issue.line());
        issue.startColumn(text.indexOf('^') + 1);
        issue.endColumn(text.lastIndexOf('^') + 2);
    }

}

File: plsql-checks-testkit/src/main/kotlin/org/sonar/plsqlopen/checks/verifier/PlSqlCheckVerifier.kt
/**
 * Z PL/SQL Analyzer
 * Copyright (C) 2015-2019 Felipe Zorzo
 * mailto:felipebzorzo AT gmail DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plsqlopen.checks.verifier

import com.google.common.base.Splitter
import com.sonar.sslr.api.Trivia
import org.assertj.core.api.Assertions.assertThat
import org.junit.Assert.fail
import org.sonar.plsqlopen.TestPlSqlVisitorRunner
import org.sonar.plsqlopen.metadata.FormsMetadata
import org.sonar.plsqlopen.squid.PlSqlAstWalker
import org.sonar.plsqlopen.symbols.DefaultTypeSolver
import org.sonar.plsqlopen.symbols.SymbolVisitor
import org.sonar.plugins.plsqlopen.api.checks.PlSqlCheck
import java.io.File
import java.util.*

class PlSqlCheckVerifier : PlSqlCheck() {

    private val expectedIssues = ArrayList<TestIssue>()

    override fun visitComment(trivia: Trivia, content: String) {
        val text = content.trim { it <= ' ' }
        val marker = "Noncompliant"

        if (text.startsWith(marker)) {
            var issueLine = trivia.token.line
            var paramsAndMessage = text.substring(marker.length).trim { it <= ' ' }

            if (paramsAndMessage.startsWith("@")) {
                val spaceSplit = paramsAndMessage.split("[\\s\\[{]".toRegex(), 2).toTypedArray()

                val shiftValue = spaceSplit[0]

                if (shiftValue[1] != '+' && shiftValue[1] != '-') {
                    fail("Use only '@+N' or '@-N' to shifts messages.")
                }

                issueLine += Integer.valueOf(shiftValue.substring(1))
                paramsAndMessage = if (spaceSplit.size > 1) spaceSplit[1] else ""
            }

            val issue = TestIssue(null, issueLine)

            if (paramsAndMessage.startsWith("{{")) {
                val endIndex = paramsAndMessage.indexOf("}}")
                val message = paramsAndMessage.substring(2, endIndex)
                issue.message = message
                paramsAndMessage = paramsAndMessage.substring(endIndex + 2).trim { it <= ' ' }
            }

            if (paramsAndMessage.startsWith("[[")) {
                val endIndex = paramsAndMessage.indexOf("]]")
                addParams(issue, paramsAndMessage.substring(2, endIndex))
            }

            expectedIssues.add(issue)

        } else if (text.startsWith("^")) {
            addPreciseLocation(trivia)
        }
    }

    private fun addPreciseLocation(trivia: Trivia) {
        val token = trivia.token
        val line = token.line
        val text = token.value
        if (token.column > 1) {
            throw IllegalStateException(
                    "Line $line: comments asserting a precise location should start at column 1")
        }
        val missingAssertionMessage = String.format(
                "Invalid test file: a precise location is provided at line %s but no issue is asserted at line %s",
                line, line - 1)
        if (expectedIssues.isEmpty()) {
            throw IllegalStateException(missingAssertionMessage)
        }
        val issue = expectedIssues[expectedIssues.size - 1]
        if (issue.line != line - 1) {
            throw IllegalStateException(missingAssertionMessage)
        }
        issue.endLine = issue.line
        issue.startColumn = text.indexOf('^') + 1
        issue.endColumn = text.lastIndexOf('^') + 2
    }

    companion object {

        @JvmStatic
        fun scanFileForIssues(file: File, metadata: FormsMetadata?, check: PlSqlCheck): List<PreciseIssue> {
            val walker = PlSqlAstWalker(Arrays.asList(SymbolVisitor(DefaultTypeSolver()), check))
            walker.walk(TestPlSqlVisitorRunner.createContext(file, metadata))
            return check.issues()
        }

        @JvmStatic
        @JvmOverloads
        fun verify(path: String, check: PlSqlCheck, metadata: FormsMetadata? = null) {
            val verifier = PlSqlCheckVerifier()
            val file = File(path)
            TestPlSqlVisitorRunner.scanFile(file, metadata, verifier)

            val actualIssues = getActualIssues(file, metadata, check)
            val expectedIssues = verifier.expectedIssues.sortedBy { it.line }

            for (expected in expectedIssues) {
                if (actualIssues.hasNext()) {
                    verifyIssue(expected, actualIssues.next())
                } else {
                    throw AssertionError("Missing issue at line " + expected.line)
                }
            }

            if (actualIssues.hasNext()) {
                val issue = actualIssues.next()
                throw AssertionError(
                        "Unexpected issue at line " + line(issue) + ": \"" + issue.primaryLocation().message() + "\"")
            }

        }

        private fun verifyIssue(expected: TestIssue, actual: PreciseIssue) {
            if (line(actual) > expected.line) {
                fail("Missing issue at line " + expected.line)
            }
            if (line(actual) < expected.line) {
                fail("Unexpected issue at line " + line(actual) + ": \"" + actual.primaryLocation().message() + "\"")
            }
            expected.message?.let {
                assertThat(actual.primaryLocation().message()).`as`("Bad message at line " + expected.line)
                        .isEqualTo(it)
            }
            expected.effortToFix?.let {
                assertThat(actual.cost()).`as`("Bad effortToFix at line " + expected.line)
                        .isEqualTo(it)
            }
            expected.startColumn?.let {
                assertThat(actual.primaryLocation().startLineOffset() + 1).`as`("Bad start column at line " + expected.line)
                        .isEqualTo(it)
            }
            expected.endColumn?.let {
                assertThat(actual.primaryLocation().endLineOffset() + 1).`as`("Bad end column at line " + expected.line)
                        .isEqualTo(it)
            }
            expected.endLine?.let {
                assertThat(actual.primaryLocation().endLine()).`as`("Bad end line at line " + expected.line)
                        .isEqualTo(it)
            }
            expected.secondaryLines?.let {
                assertThat(secondary(actual)).`as`("Bad secondary locations at line " + expected.line)
                        .isEqualTo(it)
            }
        }

        private fun secondary(issue: PreciseIssue): List<Int> {
            val result = ArrayList<Int>()

            for (issueLocation in issue.secondaryLocations()) {
                result.add(issueLocation.startLine())
            }

            return result.sorted()
        }

        private fun getActualIssues(file: File, metadata: FormsMetadata?, check: PlSqlCheck): Iterator<PreciseIssue> {
            val issues = scanFileForIssues(file, metadata, check)
            val sortedIssues = issues.sortedBy { it.primaryLocation().startLine() }
            return sortedIssues.iterator()
        }

        private fun line(issue: PreciseIssue): Int {
            return issue.primaryLocation().startLine()
        }

        private fun addParams(issue: TestIssue, params: String) {
            for (param in Splitter.on(';').split(params)) {
                val equalIndex = param.indexOf('=')
                if (equalIndex == -1) {
                    throw IllegalStateException("Invalid param at line 1: $param")
                }
                val name = param.substring(0, equalIndex)
                val value = param.substring(equalIndex + 1)

                when {
                    "effortToFix".equals(name, ignoreCase = true) -> issue.effortToFix = Integer.valueOf(value)
                    "sc".equals(name, ignoreCase = true) -> issue.startColumn = Integer.valueOf(value)
                    "ec".equals(name, ignoreCase = true) -> issue.endColumn = Integer.valueOf(value)
                    "el".equals(name, ignoreCase = true) -> issue.endLine = lineValue(issue.line, value)
                    "secondary".equals(name, ignoreCase = true) -> addSecondaryLines(issue, value)
                    else -> throw IllegalStateException("Invalid param at line 1: $name")
                }
            }
        }

        private fun addSecondaryLines(issue: TestIssue, value: String) {
            val secondaryLines = ArrayList<Int>()
            if ("" != value) {
                for (secondary in Splitter.on(',').split(value)) {
                    secondaryLines.add(lineValue(issue.line, secondary))
                }
            }
            issue.secondaryLines = secondaryLines
        }

        private fun lineValue(baseLine: Int, shift: String): Int {
            if (shift.startsWith("+")) {
                return baseLine + Integer.valueOf(shift.substring(1))
            }
            return if (shift.startsWith("-")) {
                baseLine - Integer.valueOf(shift.substring(1))
            } else Integer.valueOf(shift)
        }
    }

}
